### DISCUSSION TITLE
Initial pipeline config_ConfigureExports run after scope config failing

### DISCUSSION THREAD

**Discussion by lundejd** [microsoft/finops-toolkit#817] [General]
I went to deploy a fresh version .4 today and everything deployed fine, but when I configured for my EA scope I see the following error:

Operation on target ForEach Export Scope failed: The function 'length' expects its parameter to be an array or a string. The provided value is of type 'Object'.

![Screenshot 2024-07-12 at 7 16 27 AM](https://github.com/user-attachments/assets/1d11438a-83d4-4ccf-ac1e-dccf6b307826)

I do notice the .json config file scope is different than .3 (there are no [] characters).

Also there is no Author > Pipelines > msexports_backfill pipeline. I wanted to do it this way vs. doing manual monthly backfill exports like before. Is this just not ready yet?

**Comment by arthurclares** [comment]
@lundejd sorry for the delay on this one, John. 
For the first issue with the configureExports pipeline, can you double check if your config file looks like this?
`{
  "$schema": "https://aka.ms/finops/hubs/settings-schema",
  "type": "HubInstance",
  "version": "0.4",
  "learnMore": "https://aka.ms/finops/hubs",
"scopes": [
  {
    "scope": "xxxxx"
  }
],
  "retention": {
    "msexports": {
      "days": 0
    },
    "ingestion": {
      "months": 13
    }
  }
}
`
About the second question, there is a mistake on the docs and the pipeline name is wrong, it should say "config_RunBackfill" instead of msexports_backfill. Can you try to run this pipeline instead?
Just make sure you have assigned the correct permissions to the Data Factory managed identity at the correct scope (subscription or billing account)

**Reply by soyroberto** [reply]
my file does look like this: {
  "$schema": "https://aka.ms/finops/hubs/settings-schema",
  "type": "HubInstance",
  "version": "0.4",
  "learnMore": "https://aka.ms/finops/hubs",
  "scopes": {
    "scope": "/providers/Microsoft.Billing/billingAccounts/xxxxxxxx/"
  },
  "retention": {
    "msexports": {
      "days": 0
    },
    "ingestion": {
      "months": 13
    }
  }
}
I've tried to run using the date format yyyy-mm-dd both with '' and without - no luck
but doesn't run
![Screenshot 2024-08-27 at 16 13 46](https://github.com/user-attachments/assets/378ecfbb-965c-4e9f-b139-79ccc355004a)
---

**Reply by flanakin** [reply]
@soyroberto Can you update settings.json to use:
```json
"scopes": [{
    "scope": "/providers/Microsoft.Billing/billingAccounts/xxxxxxxx/"
}],
```
You're missing brackets around the curly braces. Also let us know if you updated this manually or via a tool/script. I just want to make sure we have the right guidance and bugs are resolved.

**Reply by soyroberto** [reply]
Thanks for the callout, modified it to:
![Screenshot 2024-08-29 at 10 28 31](https://github.com/user-attachments/assets/0c3d9357-068a-4963-a984-bfd67ad1ba1c)
I'm doing the configuration of this file manually using VScode. copy/pasting azure portal-VScode-Azureportal

**Reply by flanakin** [reply]
I see the problem. For some reason, the settings.json file is created without the brackets. I made a change in 0.5 which will handle it with or without the brackets to avoid this going forward.

Thanks for sharing the details and, again, my apologies for the inconvenience here.

**Comment by saleemmy** [comment]
I am having the same error.

**Reply by flanakin** [reply]
I wonder if you're running into the same issue as @soyroberto. Make sure the `scopes` property in settings.json is an array that is surrounded by `[` and `]`.

**Comment by soyroberto** [comment]
I'm confused documentation says: '**FinOps hubs leverage Cost Management exports to obtain cost data.** . am I supposed to be dropping the files myself in the ingestion folder? exported from the Cost management? 
+
Step 1: Verify Cost Management export
Go to Cost Management exports and make sure the export status is “Successful”.

Sources:
https://microsoft.github.io/finops-toolkit/hubs/data
https://microsoft.github.io/finops-toolkit/help/troubleshoot#-validate-your-finops-hub-deployment

Would this be the reason the DF run fails? Isn't the DF supposed to generated the cost files itself?

**Reply by soyroberto** [reply]
But the same documentation says: ' 

- **Configure managed exports Managed exports allow FinOps hubs to setup and maintain Cost Management exports for you'**

- **For the most seamless experience, we recommend allowing FinOps hubs to manage exports for you when possible. This is the simplest to setup and requires the least effort to maintain over time.**
Quite confusing

**Reply by flanakin** [reply]
Configure exports into the `msexports` container. The ADF pipeline transforms the data to parquet and moves it to the `ingestion` container. If you give your hub access to your scopes, it can manage the exports for you, but the exports are the way we get cost data.

**Reply by soyroberto** [reply]
We have permissions to the principal at this level shown in the picture:
![image](https://github.com/user-attachments/assets/28a8cb90-dec6-44a8-918b-8c8206104fb7)
how can i confirm it does generate a new export every month?

**Reply by flanakin** [reply]
When you add a scope, exports are created. Cost Management runs the exports on a schedule automatically.

**Reply by soyroberto** [reply]
I think we finally it got it working, 
1st: We configured an export using the FOCUS format and I assume this trigger the DF, would like to see where that logic sits generated a .parquet file, see pictures:

![Screenshot 2024-09-03 at 15 45 59](https://github.com/user-attachments/assets/e4559a38-7241-49eb-af49-dd9f65ba9d38)
![Screenshot 2024-09-03 at 15 44 29](https://github.com/user-attachments/assets/58be367e-f219-417b-8f16-af47a913f57a)

![Screenshot 2024-09-03 at 15 41 51](https://github.com/user-attachments/assets/ebbcde90-dd8d-42e1-b7f7-eaaa69e10b8f)

Would go to PBI and see how that goes

**Comment by saleemmy** [comment]
I updated the settings.json file with the scope. Now I get this error Operation on target ForEach Export Scope failed: Activity failed because an inner activity failed; Inner activity name: Get exports for scope, Error: {"error":{"code":"RBACAccessDenied","message":"The client does not have authorization to perform action. Request ID: NoHttpContext"}}


{
  "$schema": "https://aka.ms/finops/hubs/settings-schema",
  "type": "HubInstance",
  "version": "0.4",
  "learnMore": "https://aka.ms/finops/hubs",
  "scopes": [{
    "scope": 
		"/providers/Microsoft.Billing/billingAccounts/xxxxxxxxxxxxx"
  }],
  "retention": {
    "ingestion": {
      "months": 13
    },
    "msexports": {
      "days": 0
    }
  }
}

**Reply by flanakin** [reply]
This means the ADF managed identity does not have access to create exports for the billing account. Did you give it access? 

Email ftk-support@ms and we can setup time to make sure this is working for you.

**Comment by soyroberto** [comment]
I got it working, Once all settles, I'd be planning to write a blog about, still I have a key issue, I configured the report to cover 1 year of costs as specified here: https://microsoft.github.io/finops-toolkit/power-bi/setup#-setup-your-first-report 

**Change RangeStart and RangeEnd to the desired start/end dates for your report**

Thing is, it only brings 1 month of data (august). Also, 1 yr of data is 1.7M less than the 2M warning specified.

See images below
![Screenshot 2024-09-04 at 16 08 00](https://github.com/user-attachments/assets/2535ae61-23f2-43eb-9403-a6a2cf85e73b)

![Screenshot 2024-09-04 at 16 07 38](https://github.com/user-attachments/assets/82f0c19e-82a4-4f80-8d50-7e38da8317fe)

I tried setting the dates to:
1/1/2024
1/9/2024 
9 months and it keeps showing only august, i cannot even select dates, thoughts?
![Screenshot 2024-09-04 at 17 07 47](https://github.com/user-attachments/assets/d1cfa295-c9fa-4822-bfc5-4ea265b751f7)

**Reply by soyroberto** [reply]
I just reviewed the export settings there's no way to export 12 months, that's why I only see 1 month data, **wonder how do I get 12 months of data**?

**Reply by flanakin** [reply]
From the UI, you can open the export details and export one month at a time for the last 12 months.

From PowerShell, you can go back 7 years. See https://aka.ms/ftk/Start-FinOpsCostExport

**Comment by MS-mikeo** [comment]
@flanakin - I ran into this issue with a customer today.  We deployed .5 and still had to add the [ ] within the scopes section to get the pipeline to run.

@arthurclares - It looks like maybe the pipeline run name changed even from your comment earlier in this issue.   The documentation needs to be updated withe pipeline "config_BackfillData" when configuring the managed exports.

**Reply by flanakin** [reply]
Sorry, missed this. We investigated this further and discovered there were a few other places that need to be updated to resolve the [ ] issue. We fixed those in 0.6.

Side note: We changed the names of these pipelines to hopefully be clearer. Docs have also been updated accordingly, so they should be aligned. Let us know if you spot any issues.

**Comment by lundejd** [comment]
One thing I realized after having failure after failure is that the export script seems to only take a single month per run, not a whole range of months (similar to how Cost Management manual back-fill jobs are done) I may have missed it but this is nowhere in the documentation, nor is the format of the start and end date strings so I had to do a lot of trial and error to find something that works.

StartDate: 2024-08-01
EndDate: 2024-08-31

Seems to work just fine, but 20240801 and 20240831 or other combos will generate an error.

**Reply by soyroberto** [reply]
I made a similar comment about it - https://github.com/microsoft/finops-toolkit/discussions/817#discussioncomment-10551605 . I still have the question, how can I visualize more than 1 month of data?

**Reply by flanakin** [reply]
Ah, yes. I noticed the pipeline only supports one month at a time as well. We need to update that to scale better. For now, I would suggest trying the [Start-FinOpsCostExport command](https://aka.ms/ftk/Start-FinOpsCostExport). If it'd help, we could create a template to deploy that to Azure Automation.

**Reply by flanakin** [reply]
> I made a similar comment about it - [#817 (reply in thread)](https://github.com/microsoft/finops-toolkit/discussions/817#discussioncomment-10551605) . I still have the question, how can I visualize more than 1 month of data?

@soyroberto 

1. Export the months you need from Cost Management
2. Verify the data was pushed to the msexports container in storage
3. Verify ADF moved the data to the ingestion container in storage
4. Clear RangeStart/RangeEnd parameters to pull all data from storage (or set Number of Months in 0.5 or 0.6 reports)
5. Make sure the date picker in the report (top-left) is covering all the months you want

**Comment by soyroberto** [comment]
Running **Start-FinOpsCostExport** is where the key of exporting data is. There's no way to do the same on the webui. The webUI will work for a month's export, afterwards the PScommands must be run. 

If they're run successfully the structure on the storage account will look like + the data factory:
![Screenshot 2024-10-08 at 15 05 34](https://github.com/user-attachments/assets/f59bce1a-7873-4a0a-8c07-e8de592f5790)
![Screenshot 2024-10-08 at 15 21 45](https://github.com/user-attachments/assets/667639a0-7a5e-46a2-afc3-ad654e202fc1)
![Screenshot 2024-10-08 at 15 21 55](https://github.com/user-attachments/assets/43c47f31-4c75-4398-a876-37df627b562e)

Once the data is updated, if needed refresh the PowerBI reports. 

@flanakin **RE: If it'd help, we could create a template to deploy that to Azure Automation.** This would be fantastic. Thanks

**Reply by flanakin** [reply]
Can you [create an issue](https://aka.ms/ftk/ideas) to track that request?

**Comment by RamNadal** [comment]
Hello @flanakin , i tried to chagne the scope under the settings.json to subscriptions and got the below error

Operation on target ForEach Export Scope failed: Activity failed because an inner activity failed; Inner activity name: Create or update open month focus export, Error: {"error":{"code":"RBACAccessDenied","message":"The client does not have authorization to perform action. Request ID: dc725b8a-0af9-4886-b244-0aa69189256d"}}

**Reply by flanakin** [reply]
This sounds like the Data Factory managed identity does not have access to the subscription you're adding. Please refer to the first step in the [configure scopes doc](https://microsoft.github.io/finops-toolkit/hubs/configure#-configure-managed-exports).

**Comment by RamNadal** [comment]
@flanakin  When i chang ethe scope to my billing profile ..i get the belwo error message

Operation on target ForEach Export Scope failed: Activity failed because an inner activity failed; Inner activity name: Create or update open month focus export, Error: {"error":{"code":"MissingSubscription","message":"The request did not have a subscription or a valid tenant level resource provider."}}

**Reply by flanakin** [reply]
This sounds like the scope ID may be invalid. Make sure subscriptions are added using the scope format as `/subscriptions/############-####-####-####-########`.
